state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -957,8 +1025,115 @@ pub mod pallet {
 					debug_assert!(false, "Swap is not completed yet!");
 				}
 			}
+		}
 
-			Ok(())
+		// Returns swaps that have not expired yet and should be scheduled for retry.
+		fn refund_any_expired_swaps(
+			swaps: Vec<Swap>,
+			current_block: BlockNumberFor<T>,
+		) -> Vec<Swap> {
+			let (swaps_to_retry, swaps_to_refund): (Vec<Swap>, Vec<Swap>) =
+				swaps.into_iter().partition(|swap| {
+					swap.refund_params
+						.as_ref()
+						.map_or(true, |params| current_block < params.refund_block.into())
+				});
+
+			for swap in swaps_to_refund {
+				let destination_address = swap.refund_params.unwrap().refund_address;
+				match T::EgressHandler::schedule_egress(
+					swap.from,
+					swap.input_amount,
+					destination_address,
+					None,
+				) {
+					Ok(ScheduledEgressDetails { egress_id, egress_amount, fee_withheld }) => {
+						Self::deposit_event(Event::<T>::RefundEgressScheduled {
+							swap_id: swap.swap_id,
+							egress_id,
+							amount: egress_amount,
+							egress_fee: fee_withheld,
+						});
+					},
+					Err(err) => {
+						Self::deposit_event(Event::<T>::RefundEgressIgnored {
+							swap_id: swap.swap_id,
+							asset: swap.from,
+							amount: swap.input_amount,
+							reason: err.into(),
+						});
+					},
+				}
+			}
+
+			swaps_to_retry
+		}
+
+		fn process_swaps_for_block(
+			current_block: BlockNumberFor<T>,
+			block_to_process: BlockNumberFor<T>,
+		) -> Result<(), ()> {
+			let mut swaps_to_execute = SwapQueue::<T>::take(block_to_process);
+
+			loop {
+				if swaps_to_execute.is_empty() {
+					return Ok(())
+				}
+
+				match Self::execute_batch(swaps_to_execute.clone()) {
+					Ok(successful_swaps) => {
+						Self::process_swap_outcomes(&successful_swaps);
+						// Nothing to do here, all swaps are processed for block
+						return Ok(())
+					},
+					Err(err) => {
+						let (successful_swaps, failed_swaps) = {
+							match err {
+								BatchExecutionError::PriceLimitHit {
+									successful_swaps,
+									failed_swaps,
+								} => (successful_swaps, failed_swaps),
+								BatchExecutionError::SwapLegFailed { asset, direction, amount } => {
+									Self::deposit_event(Event::<T>::BatchSwapFailed {
+										asset,
+										direction,
+										amount,
+									});
+									(vec![], swaps_to_execute)
+								},
+								BatchExecutionError::DispatchError { error } => {
+									log::error!(
+										"Failed to execute swap batch at block {:?}: {:?}",
+										block_to_process,
+										error
+									);
+									(vec![], swaps_to_execute)
+								},
+							}
+						};
+
+						let swaps_to_retry =
+							Self::refund_any_expired_swaps(failed_swaps, current_block);
```

Comment:
I feel like this function just be expressed a bit more neatly using `extract_if`.

Reply:
I refactored the code a bit (by inlining `refund_any_expired_swaps`) so that params doesn't need to be unwrapped, which means `extract_if` is no longer needed.